core/src/avm2/traits.rs

```diff
@@ -29,6 +31,58 @@ bitflags! {
     }
 }
 
+// Represents a single key-value pair for a trait metadata.
+#[derive(Clone, Collect, Debug, Eq, PartialEq)]
+#[collect(no_drop)]
+pub struct TraitMetadataItem<'gc> {
+    pub key: AvmString<'gc>,
+    pub value: AvmString<'gc>,
+}
+
+// Represents a single metadata item for a trait.
+#[derive(Clone, Collect, Debug, Eq, PartialEq)]
+#[collect(no_drop)]
+pub struct TraitMetadata<'gc> {
+    pub name: AvmString<'gc>,
+    pub items: Vec<TraitMetadataItem<'gc>>,
+}
+
+impl<'gc> TraitMetadata<'gc> {
+    // Represents this TraitMetadata as an XML string, as defined in describeType
+    pub fn as_xml_string(&self, mc: MutationContext<'gc, '_>) -> AvmString<'gc> {
+        let mut xml_string = AvmString::new_utf8(mc, "");
+        if self.items.is_empty() {
+            // This was in the form of [metadata]
+            xml_string = AvmString::concat(
+                mc,
+                xml_string,
+                AvmString::new_utf8(mc, format!("<metadata name=\"{}\"/>", self.name)),
```

Comment:
most/all of this should probably be in `utils.rs` together with the other DescribeType-related XML string creation code.